### DISCUSSION TITLE
What build files `c3c clean` should supposedly clean?

### DISCUSSION THREAD

**Discussion by stefanos82** [c3lang/c3c#1256] [Q&A]
I run `c3c init mydefer` and in `src`, I copied the `defer` example from https://c3-lang.org/references/docs/examples/ and tweaked its module name to avoid compilation errors due to reserved keyword `defer`.

Command `c3c build` worked as expected, the `c3c run` produced the following output:

```txt
Program linked to executable 'build/mydefer'.
Launching ./build/mydefer...
A
BA
B!A
Program completed with exit code 0.
```

So far, so good!

Now, I wanted to clean the entire project, therefore I run `c3c clean`, but nothing happened; I can still see the `mydefer` binary in `build` folder.

Am I doing something wrong?

My version is

```
C3 Compiler Version:       0.6.1 (prerelease)
Installed directory:       /PortableApps/c3c/bin/
LLVM version:              16.0.6
LLVM default target:       x86_64-pc-linux-gnu
```

**Comment by lerno** [comment]
There's actually an unclosed issue for this: #456 

"c3c clean" will currently clean up the object files it created. But it won't remove the executable. The question is "what kind of behaviour do we expect?". When there is a project, then it's fairly straightforward: clean the build directory. But then, is the exe in the (temporary) build directory or elsewhere?

Unlike some build systems, where you have to be in the exact top build directory to invoke "build", you can do it further down in a project directory. So then where should the executable and temporary build files end up relative to the project file and the current location?

If instead "compile" / "compile-run" is used, the situation is even more difficult. We want to place the temporary files in a sub folder, but the user might already be using any directory we might create by default - unless we generate a name. But if we generate a name, how will the run of "clean" know that this particular directory is to be deleted? And if there is no default sub folder, how do we know that we're just deleting the object files that was last created?

These are not HARD questions, but some convention has to be enforced, and it's a convention that should be similar to other build scripts / build systems.

So the bottom line is: what behaviour is preferred and expected?

**Comment by stefanos82** [comment]
To me the concept of `c3c clean` should be more or less identical to that of Rust Cargo's: `remove the target directory`.

When I execute the aforementioned command, it means I know **exactly** what I'm doing, thus the need of execution of this command; I want to clean up everything and rebuild the entire project from start.

**Reply by lerno** [reply]
There should probably be additional limitations, since it's possible to redirect the build directory. So let's say you set the build directory to be `./`, in this case the "clean" would delete everything in the project directory. Something that probably would not be popular. So the question is what to do for non-standard build directories.

**Comment by stefanos82** [comment]
Seems like I have answered my own question after going through an investigation inside Cargo's produced output:

![image](https://github.com/user-attachments/assets/5f3d19f7-55df-4d07-a45f-f4527e2e71e1)

Maybe this is the concept that could help C3 to simplify things even further?

Food for thought :thinking:

**Comment by stefanos82** [comment]
Seems like what I wanted, can be done with `c3c run --run-once`.

The only "issue" I have is the `Program linked to executable 'build/hello_world'.` and `Program completed with exit code 0.` messages; I would like to silent them.

**Reply by lerno** [reply]
File an issue for a "--silent" flag?

**Reply by stefanos82** [reply]
You mean to open a new request ticket?

**Reply by stefanos82** [reply]
@lerno Done; please see #1601 

Thank you very much.
